lib/services/android-debug-service.ts

```diff
@@ -43,14 +39,60 @@ class AndroidDebugService implements IDebugService {
 			: this.debugOnDevice();
 	}
 
-	public debugOnEmulator(): IFuture<void> {
+	private debugOnEmulator(): IFuture<void> {
 		return (() => {
 			this.$platformService.deployOnEmulator(this.platform).wait();
 			this.debugOnDevice().wait();
 		}).future<void>()();
 	}
 
-	public debugOnDevice(): IFuture<void> {
+	private isPortAvailable(candidatePort: number): IFuture<boolean> {
+		let future = new Future<boolean>();
+		let server = net.createServer();
+		server.on("error", (err: Error) => {
+			future.return(false);
+		});
+		server.listen(candidatePort, (err: Error) => {
+			server.once("close", () => {
+				future.return(true);
+			});
+			server.close();
+		});
+
+		return future;
+	}
+
+	private getForwardedLocalDebugPortForPackageName(deviceId: string, packageName: string): IFuture<number> {
+		return (() => {
+			let port = -1;
+			let forwardsResult = this.device.adb.executeCommand(["forward", "--list"]).wait();
+
+			//matches 123a188909e6czzc tcp:40001 localabstract:org.nativescript.testUnixSockets-debug
+			let regexp = new RegExp(`(${deviceId} tcp:)([\\d])+(?= localabstract:${packageName}-debug)`, "g");
```

Comment:
I believe that this RegExp could be rewritten as follows:
```Typescript
new RegExp(`(?:${deviceId} tcp:)([\\d]+)(?= localabstract:${packageName}-debug)`, "g");
```

The benefits would be that you could execute `regexp.exec(forwardsResult)` on the line below instead of `forwardsResult.match(regexp)`. That way the result port would simply be `match[1]` instead of `eInt(match[0].substring(match[0].length - 5));`

Example:
```TypeScript
let deviceId = "123a188909e6czzc";
let packageName = "org.nativescript.testUnixSockets";
let forwardsResult = "123a188909e6czzc tcp:40001 localabstract:org.nativescript.testUnixSockets-debug";

let regexp = new RegExp(`(?:${deviceId} tcp:)([\\d]+)(?= localabstract:${packageName}-debug)`, "g");
let match = regexp.exec(forwardsResult);
console.log(match);
// output
/* [ '123a188909e6czzc tcp:40001',
  '40001',
  index: 0,
  input: '123a188909e6czzc tcp:40001 localabstract:org.nativescript.testUnixSockets-debug' ]
*/
```